syntax = "proto3";

package larcs.msgs;

import "common.proto";

// PC → ESP32 指令
message ESP32Command {
  Header header = 1;
  
  enum Mode {
    DISARM = 0;   // モータ無効
    ARM = 1;      // モータ有効
    ESTOP = 2;    // 緊急停止
  }
  Mode mode = 2;
  
  // 車輪速度目標 [RPM] (最大4輪対応)
  repeated double wheel_rpm_target = 3;
  
  // タイムアウト [ms] (この時間内に次のコマンドが来なければ停止)
  uint32 ttl_ms = 4;
}

// ESP32 → PC 状態
message ESP32State {
  Header header = 1;
  
  // エンコーダ積算値 [ticks]
  repeated int64 encoder_ticks = 2;
  
  // 車輪速度推定 [RPM]
  repeated double wheel_rpm = 3;
  
  // バッテリ電圧 [V]
  double battery_voltage = 4;
  
  // E-stopボタン押下状態
  bool estop_pressed = 5;
  
  // ESP32起動時間 [ms]
  uint32 uptime_ms = 6;
  
  // モータドライバエラー
  repeated string errors = 7;
}
